### DISCUSSION TITLE
Support Azure OpenAi deployments

### DISCUSSION THREAD

**Discussion by edubrigham** [n4ze3m/dialoqbase#294] [Feature request]
Due to GDPR constraints, it would be great to use Azure Openai deployments.
In Europe we're not allowed to use OpenAI gpt models for public consumption
because they are hosted on American servers.

**Comment by n4ze3m** [comment]
Hey, Dialoqbase support. For the Azure API, you need to add the following variables to .env:

- AZURE_OPENAI_API_KEY
- AZURE_OPENAI_API_VERSION
- AZURE_OPENAI_API_INSTANCE_NAME
- AZURE_OPENAI_API_DEPLOYMENT_NAME

Let me know if you have any issues

**Reply by edubrigham** [reply]
Thank you so much for your time and assistance!
I have added these variables to the .env. file and restarted the docker instance. Should I be seeing the azure deployment model in the model list? I am only seeing the default models.

**Reply by n4ze3m** [reply]
Hey, You need to manually add the model. Go to Admin > Models > Add Model, then add the OpenAI model

**Reply by edubrigham** [reply]
Thank you so much, it worked perfectly!
One small question though, the loader is failing to load any data source I give.
The docker instance outputs an embeddings model error. Do I need to use a different embedding model?:
```
error: {
2024-09-04 08:51:13 dialoqbase           |     code: 'OperationNotSupported',
2024-09-04 08:51:13 dialoqbase           |     message: 'The embeddings operation does not work with the specified model, gpt-35-turbo. Please choose different model and try again. You can learn more about which models can be used with each operation here: https://go.microsoft.com/fwlink/?linkid=2197993.'
2024-09-04 08:51:13 dialoqbase           |   },
2024-09-04 08:51:13 dialoqbase           |   code: 'OperationNotSupported',
2024-09-04 08:51:13 dialoqbase           |   param: undefined,
2024-09-04 08:51:13 dialoqbase           |   type: undefined,
2024-09-04 08:51:13 dialoqbase           |   attemptNumber: 1,
2024-09-04 08:51:13 dialoqbase           |   retriesLeft: 6
2024-09-04 08:51:13 dialoqbase           | }
```

**Reply by n4ze3m** [reply]
Hey, you need to add the embedding model on the same page as the chat model. Choose the embedding tag and add the model.

**Reply by edubrigham** [reply]
I'm very sorry to bother you again. unfortunately I still didn't manage to get the chatbot to connect to my azure model
You said that I need to add the following variables to .env:

AZURE_OPENAI_API_KEY
AZURE_OPENAI_API_VERSION
AZURE_OPENAI_API_INSTANCE_NAME
AZURE_OPENAI_API_DEPLOYMENT_NAME

as per screenshots, I added the following entries to the .env file in the Docker folder:
AZURE_OPENAI_API_DEPLOYMENT_NAME="gpt-4"
AZURE_OPENAI_API_INSTANCE_NAME="azureopenai-francecentral-instance"
AZURE_OPENAI_API_KEY="my_azure_openai_key"
AZURE_OPENAI_API_VERSION="2024-05-01-preview"

and deployed a gpt-4 chat model and ada-002 embeddings model, but I still get the error message:
code: 'OperationNotSupported',
2024-09-04 10:19:22 dialoqbase           |     message: 'The embeddings operation does not work with the specified model, gpt-4. Please choose different model and try again. You can learn more about which models can be used with each operation here: https://go.microsoft.com/fwlink/?linkid=2197993.'


What am I doing wrong?
Thanks a million for your help!

<img width="566" alt="Screenshot 2024-09-04 at 10 57 08" src="https://github.com/user-attachments/assets/a0714f77-9b93-4887-87f0-6c1be463f486">
<img width="531" alt="Screenshot 2024-09-04 at 11 12 33" src="https://github.com/user-attachments/assets/8e05b307-7888-4efe-acc5-89a374d528db">
<img width="1499" alt="Screenshot 2024-09-04 at 11 09 15" src="https://github.com/user-attachments/assets/328001c7-8988-4c2c-96a9-f8b058e13341">
<img width="1298" alt="Screenshot 2024-09-04 at 11 09 05" src="https://github.com/user-attachments/assets/5b4d5a14-48b0-4818-9a33-22423ec35d0f">
<img width="787" alt="Screenshot 2024-09-04 at 11 04 07" src="https://github.com/user-attachments/assets/21c94a13-d6ac-4ed1-ad61-d5a1d536b3f7">

**Reply by n4ze3m** [reply]
Hey, it's okay! 😊 I can assist you. I think you need to add the `AZURE_OPENAI_API_EMBEDDINGS_DEPLOYMENT_NAME` variable for the embedding model, just like in chat. Let me know if it works!

**Reply by edubrigham** [reply]
Thank you for all the help, I used the following .env variables and it worked:

AZURE_OPENAI_API_DEPLOYMENT_NAME="gpt-4"
AZURE_OPENAI_API_INSTANCE_NAME="azureopenai-francecentral-instance"
AZURE_OPENAI_API_KEY="my_azure_key"
AZURE_OPENAI_API_VERSION="2024-05-01-preview"
AZURE_OPENAI_API_EMBEDDINGS_DEPLOYMENT_NAME="text-embedding-ada-002"
AZURE_OPENAI_API_EMBEDDINGS_INSTANCE_NAME="azureopenai-francecentral-instance"
AZURE_OPENAI_API_EMBEDDINGS_KEY="my_azure_key"


but this setup leads me to a query: if I need to hardcode the deployment name for a specific model in the .env file, then does this mean I can only use this one specific model for all new RAG bots I create? I can't have multiple azure deployments/models in the .env file can I? if I want to have one bot using gpt-4o and another using gpt-4o-mini, it won't work, right? Also, is it possible to declare a different the AZURE_OPENAI_API_KEY inside the Chat Model API Key input field?



<img width="800" alt="Screenshot 2024-09-04 at 12 01 09" src="https://github.com/user-attachments/assets/ee968670-b4fb-4d21-8dec-3b7f14223776">
 


<img width="602" alt="Screenshot 2024-09-04 at 11 52 40" src="https://github.com/user-attachments/assets/537578ff-b8c5-491f-b901-4a03b376c83e">
<img width="889" alt="Screenshot 2024-09-04 at 11 51 58" src="https://github.com/user-attachments/assets/96de0b8e-82b5-4803-a9be-0c860a034a0a">

**Reply by n4ze3m** [reply]
Glad it works! I will make it easier to add models from the UI in an upcoming update. :)

**Comment by edwhiskey** [comment]
> Due to GDPR constraints, it would be great to use Azure Openai deployments. In Europe we're not allowed to use OpenAI gpt models for public consumption because they are hosted on American servers.

Hello edubrigham,

May I know how it work for you? I couldn't use the model after I added the Azure openAi key.
I tried to add the model in Admin portal but still not working.
